### PR TITLE
[dashboard][kubernetes] Dashboard CPU and memory adjustments.

### LINKED ISSUE #21353: [kubernetes][Bug] Error computing CPU usage on minikube
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Clusters

### What happened + What you expected to happen

Refer to discussion:
https://discuss.ray.io/t/logging-error-computing-cpu-usage-of-ray-kubernetes/4435

### Versions / Dependencies

Refer to discussion:
https://discuss.ray.io/t/logging-error-computing-cpu-usage-of-ray-kubernetes/4435

### Reproduction script

Refer to discussion:
https://discuss.ray.io/t/logging-error-computing-cpu-usage-of-ray-kubernetes/4435

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/21353 and
Fixes an issue that causes dashboard to read K8s CPU requests rather than resources when determining CPUs available.

In more detail, the changes are:
(1) Makes Ray's logic for determining CPU and memory compatible with cgroups v2
(2) Stops using cpu.shares to determine number of cpus available to a k8s container. That determines cpu requests, not limits. Falling back to number of available cpus or even host cpus is acceptable when cpu limits aren't set.
(3) If docker cpu counting logic concludes there are 0 CPUs available, that means there's a fractional number of cpus allocated. Round up to 1 in that case.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
   - [x] Manually tested dashboard in different circumstances.

### CODE REVIEW

**Review by ijrsvt — COMMENTED** [review]
Quick Question--does something like psutil not do this automatically?

File: python/ray/_private/utils.py [hunk 1]
```diff
@@ -456,7 +461,7 @@ def _get_docker_cpus(
     cpu_quota = None
     # See: https://bugs.openjdk.java.net/browse/JDK-8146115
     if os.path.exists(cpu_quota_file_name) and os.path.exists(
-            cpu_quota_file_name):
+            cpu_period_file_name):
```
[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/_private/utils.py:464]: psutil unfortunately reads system resources -- it doesn't look at cgroups

[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/_private/utils.py:464]: RIP!

File: python/ray/_private/utils.py [hunk 2]
```diff
@@ -465,8 +470,23 @@ def _get_docker_cpus(
...
             logger.exception("Unexpected error calculating docker cpu quota.",
                              e)
+    # Look at cpu.max for cgroups v2
+    elif os.path.exists(cpu_max_file_name):
+        try:
+            max_file = open(cpu_max_file_name).read()
+            quota_str, period_str = max_file.split()
+            if quota_str.isnumeric() and period_str.isnumeric():
+                cpu_quota = float(quota_str) / float(period_str)
+            else:
```
Comment (wuisawesome) [inline python/ray/_private/utils.py:480]: Don't think we need this case?

File: python/ray/_private/utils.py [hunk 3]
```diff
@@ -465,8 +470,23 @@ def _get_docker_cpus(
...
+                # quota_str is "max" meaning the cpu quota is unset
+                cpu_quota = None
+        except Exception:
+            logger.exception("Unexpected error calculating docker cpu quota.")
     if (cpu_quota is not None) and (cpu_quota < 0):
```
Comment (wuisawesome) [inline python/ray/_private/utils.py:485]: I think we can rephrase these in a single if statement? (also i don't think we're allowed to have fractional cpu's in the total at all?) 

```
if cpu_quota is not None and cpu_quota >= 0: 
  cpu_quota = math.ceil(cpu_quota, 1)
```

**Review by wuisawesome — APPROVED** [review]

File: python/ray/_private/utils.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/_private/utils.py:480]: You're right it's redundant.

File: python/ray/_private/utils.py (hunk 3 above)
Comment (DmitriGekhtman) [reply python/ray/_private/utils.py:485]: That would miss the functionality of setting a negative cpu_quota to None.
Btw, is it actually possible to have negative cpu quota or is this defensive programming?

Comment (wuisawesome) [reply python/ray/_private/utils.py:485]: i don't think so? or maybe there's a possibility of `-1` floating around out there? maybe ignore this comment actually, may have been an over-eager attempt to simplify this logic.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: python/ray/_private/utils.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/_private/utils.py:490]: Prefer to keep for clarity, though.

**Comment by DmitriGekhtman** [comment]
Closes https://github.com/ray-project/ray/issues/21353 and
Fixes an issue that causes dashboard to read K8s CPU requests rather than resources when determining CPUs available.

**Pull request merged**